File: go/src/relay/workers/notificationworker.go
package workers

import (
	"context"
	"log/slog"

	"github.com/drim-dev/verifying-agent-code/go/src/relay/infra"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/seams"
	amqp "github.com/rabbitmq/amqp091-go"
)

// MaxAttempts is the delivery cap before a job is dead-lettered.
const MaxAttempts = 3

// NotificationWorker consumes notify.dm with manual acks, prefetch 1: a job is acked only
// after the recorder persists its effect. A failing job is retried up to MaxAttempts, then
// dead-lettered (a final requeue:false nack routes it to the DLX → DLQ deterministically),
// so the queue keeps flowing past a poison job.
//
// The attempt cap is enforced here via x-acquired-count (the header a quorum queue stamps on
// a requeued nack), NOT by leaning on the broker's x-delivery-limit (which counts dead-letter
// republishes, not requeued nacks, so it would loop). The correct recorder treats a
// redelivered duplicate as success → ack, so a duplicate never crash-loops; the naive variant
// does not, so a redelivered duplicate dead-letters after MaxAttempts.
type NotificationWorker struct {
	conn     *amqp.Connection
	recorder seams.NotificationRecorder
	queue    string
	log      *slog.Logger
}

func NewNotificationWorker(conn *amqp.Connection, recorder seams.NotificationRecorder, queue string, log *slog.Logger) *NotificationWorker {
	return &NotificationWorker{conn: conn, recorder: recorder, queue: queue, log: log}
}

// Run consumes until ctx is cancelled.
func (w *NotificationWorker) Run(ctx context.Context) error {
	ch, err := w.conn.Channel()
	if err != nil {
		return err
	}
	defer func() { _ = ch.Close() }()
	if err := infra.DeclareNotificationQueues(ch, w.queue); err != nil {
		return err
	}
	if err := ch.Qos(1, 0, false); err != nil {
		return err
	}
	deliveries, err := ch.Consume(w.queue, "", false, false, false, false, nil)
	if err != nil {
		return err
	}

	for {
		select {
		case <-ctx.Done():
			return nil
		case delivery, ok := <-deliveries:
			if !ok {
				return nil
			}
			w.handle(ctx, delivery)
		}
	}
}

func (w *NotificationWorker) handle(ctx context.Context, delivery amqp.Delivery) {
	job, err := infra.DeserializeJob(delivery.Body)
	if err == nil {
		err = w.recorder.Record(ctx, job)
	}
	if err == nil {
		_ = delivery.Ack(false)
		return
	}
	exhausted := acquiredCount(delivery) >= MaxAttempts
	w.log.Debug("notification job failed", "attempt", acquiredCount(delivery), "exhausted", exhausted, "err", err)
	// On the final attempt nack with requeue:false → DLX → DLQ; otherwise requeue.
	_ = delivery.Nack(false, !exhausted)
}

// acquiredCount returns this delivery attempt (1-based). A quorum queue stamps
// x-acquired-count = prior acquisitions on a requeued nack (absent on first delivery).
func acquiredCount(delivery amqp.Delivery) int64 {
	if delivery.Headers == nil {
		return 1
	}
	raw, ok := delivery.Headers["x-acquired-count"]
	if !ok {
		return 1
	}
	switch v := raw.(type) {
	case int64:
		return v + 1
	case int32:
		return int64(v) + 1
	case int:
		return int64(v) + 1
	default:
		return 1
	}
}

File: python/src/relay/workers/notification_worker.py
"""The RabbitMQ notification worker.

Consumes ``notify.dm`` with manual acks, prefetch 1: a job is acked only after
the recorder persists its effect. A failing job is retried up to ``MAX_ATTEMPTS``
then dead-lettered (a final ``requeue=False`` nack routes it to the DLX → DLQ),
so the queue keeps flowing past a poison job.

The attempt cap is enforced HERE via ``x-acquired-count`` (the header a RabbitMQ
4.x quorum queue stamps on a requeued nack), NOT by leaning on the broker's
``x-delivery-limit`` (which counts dead-letter republishes, not requeued nacks,
so it would loop). The correct recorder treats a redelivered duplicate as
success → ack, so a duplicate never crash-loops; the naive variant does not, so
a redelivered duplicate dead-letters after MAX_ATTEMPTS.
"""

from __future__ import annotations

import logging
import threading

import pika

from relay import seams
from relay.infra.codecs import deserialize_job
from relay.infra.rabbit_infra import declare_notification_queues
from relay.workers import MAX_ATTEMPTS

log = logging.getLogger("relay.notification_worker")


class NotificationWorker:
    def __init__(
        self,
        params: pika.ConnectionParameters,
        recorder: seams.NotificationRecorder,
        queue: str,
    ) -> None:
        self._params = params
        self._recorder = recorder
        self._queue = queue
        self._thread: threading.Thread | None = None
        self._connection: pika.BlockingConnection | None = None
        self._stop = threading.Event()

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, name="notify-worker", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        connection = self._connection
        if connection is not None and connection.is_open:
            connection.add_callback_threadsafe(connection.close)
        if self._thread is not None:
            self._thread.join(timeout=5)

    def _run(self) -> None:
        self._connection = pika.BlockingConnection(self._params)
        channel = self._connection.channel()
        declare_notification_queues(channel, self._queue)
        channel.basic_qos(prefetch_count=1)
        for method, properties, body in channel.consume(self._queue, inactivity_timeout=0.2):
            if self._stop.is_set():
                break
            if method is None:
                continue  # inactivity tick — re-check the stop flag
            self._handle(channel, method, properties, body)
        channel.cancel()

    def _handle(self, channel, method, properties, body) -> None:
        try:
            job = deserialize_job(body)
            self._recorder.record(job)
        except Exception:
            attempt = _acquired_count(properties)
            exhausted = attempt >= MAX_ATTEMPTS
            log.debug("notification job failed; attempt=%s exhausted=%s", attempt, exhausted)
            # On the final attempt nack with requeue=False → DLX → DLQ; else requeue.
            channel.basic_nack(method.delivery_tag, requeue=not exhausted)
            return
        channel.basic_ack(method.delivery_tag)


def _acquired_count(properties: pika.BasicProperties) -> int:
    """This delivery attempt (1-based). A quorum queue stamps ``x-acquired-count``
    = prior acquisitions on a requeued nack (absent on first delivery)."""
    headers = getattr(properties, "headers", None) or {}
    raw = headers.get("x-acquired-count")
    if raw is None:
        return 1
    return int(raw) + 1
